File: Android/wallet/src/main/java/io/outblock/wallet/errors/WalletError.kt
package io.outblock.wallet.errors

import java.lang.RuntimeException

/**
 * Represents all possible errors that can occur in FlowWalletKit operations
 */
class WalletError(
    val code: Int,
    override val message: String
) : RuntimeException(message) {

    override fun toString(): String {
        return "WalletError Code: $code-$message"
    }

    companion object {
        // MARK: - General Errors
        val NoImplement = WalletError(0, "Operation or feature not implemented")
        val EmptyKeychain = WalletError(1, "No keys found in keychain")
        val EmptyKey = WalletError(2, "Key data is empty or invalid")
        val EmptySignKey = WalletError(3, "Signing key is empty or not available")

        // MARK: - Cryptographic Errors
        val UnsupportedHashAlgorithm = WalletError(4, "Hash algorithm not supported")
        val UnsupportedSignatureAlgorithm = WalletError(5, "Signature algorithm not supported")
        val InitChaChaPolyFailed = WalletError(6, "Failed to initialize ChaCha20-Poly1305")
        val InitHDWalletFailed = WalletError(7, "Failed to initialize HD wallet")
        val InitPrivateKeyFailed = WalletError(8, "Failed to initialize private key")
        val RestoreWalletFailed = WalletError(9, "Failed to restore wallet from backup")
        val InvalidSignatureAlgorithm = WalletError(10, "Invalid signature algorithm specified")
        val InvalidEVMAddress = WalletError(11, "Invalid EVM address")

        // MARK: - Authentication Errors
        val InvalidPassword = WalletError(12, "Invalid password provided")
        val InvalidPrivateKey = WalletError(13, "Invalid private key format")
        val InvalidKeyStoreJSON = WalletError(14, "Invalid KeyStore JSON format")
        val InvalidKeyStorePassword = WalletError(15, "Invalid KeyStore password")
        val SignError = WalletError(16, "Error during signing operation")
        val InitPublicKeyFailed = WalletError(17, "Failed to initialize public key")

        // MARK: - Network Errors
        val IncorrectKeyIndexerURL = WalletError(18, "Invalid key indexer URL")
        val KeyIndexerRequestFailed = WalletError(19, "Key indexer request failed")
        val DecodeKeyIndexerFailed = WalletError(20, "Failed to decode key indexer response")

        // MARK: - Storage Errors
        val LoadCacheFailed = WalletError(21, "Failed to load data from cache")
        val InvalidWalletType = WalletError(22, "Invalid wallet type for operation")

        // MARK: - Connection Errors
        val InvalidConnectionType = WalletError(23, "Invalid connection type")
        val ConnectionFailed = WalletError(24, "Failed to establish connection")
        val DisconnectionFailed = WalletError(25, "Failed to disconnect")
        val InvalidDeepLink = WalletError(26, "Invalid deep link format")
        val SessionExpired = WalletError(27, "Connection session expired")
        val InvalidSession = WalletError(28, "Invalid connection session")
        val NetworkNotSupported = WalletError(29, "Network not supported")
        val ConnectionTimeout = WalletError(30, "Connection attempt timed out")

        fun fromCode(code: Int): WalletError {
            return when (code) {
                0 -> NoImplement
                1 -> EmptyKeychain
                2 -> EmptyKey
                3 -> EmptySignKey
                4 -> UnsupportedHashAlgorithm
                5 -> UnsupportedSignatureAlgorithm
                6 -> InitChaChaPolyFailed
                7 -> InitHDWalletFailed
                8 -> InitPrivateKeyFailed
                9 -> RestoreWalletFailed
                10 -> InvalidSignatureAlgorithm
                11 -> InvalidEVMAddress
                12 -> InvalidPassword
                13 -> InvalidPrivateKey
                14 -> InvalidKeyStoreJSON
                15 -> InvalidKeyStorePassword
                16 -> SignError
                17 -> InitPublicKeyFailed
                18 -> IncorrectKeyIndexerURL
                19 -> KeyIndexerRequestFailed
                20 -> DecodeKeyIndexerFailed
                21 -> LoadCacheFailed
                22 -> InvalidWalletType
                23 -> InvalidConnectionType
                24 -> ConnectionFailed
                25 -> DisconnectionFailed
                26 -> InvalidDeepLink
                27 -> SessionExpired
                28 -> InvalidSession
                29 -> NetworkNotSupported
                30 -> ConnectionTimeout
                else -> throw IllegalArgumentException("Unknown error code: $code")
            }
        }
    }
} 

File: iOS/FlowWalletKit/Sources/WalletError.swift
//
//  File.swift
//
//
//  Created by Hao Fu on 16/1/2024.
//

/// FlowWalletKit - Error Definitions
///
/// This module defines all possible errors that can occur within the FlowWalletKit.
/// Each error case represents a specific failure scenario in the wallet operations.
///
/// The errors are categorized into several groups:
/// - General Errors: Basic operational failures
/// - Cryptographic Errors: Issues with cryptographic operations
/// - Authentication Errors: Problems with keys and signatures
/// - Network Errors: Communication failures with external services
/// - Storage Errors: Issues with data persistence
///
/// Usage Example:
/// ```swift
/// do {
///     try wallet.sign(transaction)
/// } catch WalletError.emptySignKey {
///     // Handle missing signing key
/// } catch WalletError.signError {
///     // Handle signing failure
/// }
/// ```

import Flow
import Foundation

/// Represents all possible errors that can occur in FlowWalletKit operations
///
/// This enum implements several protocols:
/// - `Error`: For Swift error handling
/// - `CaseIterable`: To enumerate all possible errors
/// - `CustomStringConvertible`: For human-readable error messages
public enum FWKError: String, Error, CaseIterable, CustomStringConvertible {
    // MARK: - General Errors
    
    /// Operation or feature not implemented
    /// Thrown when attempting to use functionality that is planned but not yet available
    case noImplement
    /// No keys found in keychain
    /// Thrown when trying to access a key that should exist in the keychain but doesn't
    case emptyKeychain
    /// Key data is empty or invalid
    /// Thrown when key data is missing or corrupted
    case emptyKey
    /// Signing key is empty or not available
    /// Thrown when attempting to sign without a valid signing key
    case emptySignKey
    
    // MARK: - Cryptographic Errors
    
    /// Hash algorithm not supported
    /// Thrown when trying to use an unsupported hashing algorithm
    case unsupportHashAlgorithm
    /// Signature algorithm not supported
    /// Thrown when trying to use an unsupported signing algorithm
    case unsupportSignatureAlgorithm
    /// Failed to initialize ChaCha20-Poly1305
    /// Thrown when encryption setup fails
    case initChaChapolyFailed
    /// Failed to initialize HD wallet
    /// Thrown when HD wallet creation fails (e.g., invalid entropy)
    case initHDWalletFailed
    /// Failed to initialize private key
    /// Thrown when private key creation fails
    case initPrivateKeyFailed
    /// Failed to restore wallet from backup
    /// Thrown when wallet restoration fails (e.g., corrupted backup)
    case restoreWalletFailed
    /// Invalid signature algorithm specified
    /// Thrown when an unknown or invalid signature algorithm is provided
    case invaildSignatureAlgorithm
    /// Invalid EVM address
    /// Thrown when an unknown or invalid EVM address is provided
    case invaildEVMAddress
    
    // MARK: - Authentication Errors
    
    /// Invalid password provided
    /// Thrown when decryption fails due to wrong password
    case invaildPassword
    /// Invalid private key format
    /// Thrown when private key data is malformed
    case invaildPrivateKey
    /// Invalid KeyStore JSON format
    /// Thrown when KeyStore JSON is malformed or missing required fields
    case invaildKeyStoreJSON
    /// Invalid KeyStore password
    /// Thrown when KeyStore decryption fails due to wrong password
    case invaildKeyStorePassword
    /// Error during signing operation
    /// Thrown when transaction or message signing fails
    case signError
    /// Failed to initialize public key
    /// Thrown when public key derivation or loading fails
    case initPublicKeyFailed
    /// Ethereum capability is unavailable for the selected key
    case unsupportedEthereumKey
    /// Unsupported or out-of-range Ethereum derivation index
    case unsupportedEthereumDerivation
    /// Invalid EVM typed data structure
    case invalidEthereumTypedData
    /// Ethereum message/digest must be 32 bytes
    case invalidEthereumMessage
    /// Malformed Ethereum signature blob
    case invalidEthereumSignature
    /// Invalid or malformed Ethereum derivation path definition
    case invalidEthereumDerivationPath
    
    case failedPassSecurityCheck
    
    // MARK: - Network Errors
    
    /// Invalid key indexer URL
    /// Thrown when the key indexer service URL is malformed
    case incorrectKeyIndexerURL
    /// Key indexer request failed
    /// Thrown when communication with key indexer service fails
    case keyIndexerRequestFailed
    /// Failed to decode key indexer response
    /// Thrown when key indexer response is invalid or unexpected
    case decodeKeyIndexerFailed
    
    // MARK: - Storage Errors
    
    /// Failed to load data from cache
    /// Thrown when cached data cannot be retrieved or is corrupted
    case loadCacheFailed
    /// Invalid wallet type for operation
    /// Thrown when attempting an operation not supported by the wallet type
    case invaildWalletType
    
    case cacheDecodeFailed
    
    case emptyCreatedAddress
    
    case noImplementError

    /// Returns the numeric error code for the error case
    /// This can be used for error tracking and analytics
    var errorCode: Int {
        FWKError.allCases.firstIndex(of: self) ?? -1
    }

    /// Returns a human-readable description of the error
    /// Format: "WalletError Code: {code}-{rawValue}"
    public var description: String {
        "\(type(of: self)) Code: \(errorCode)-\(self.rawValue)"
    }
}
